### PR TITLE
Multistage build to create an image for running the webapp 54MB compressed

### PR DESCRIPTION
**Pull request by mauritsvdvijgh**
Based on: https://github.com/Algram/ytdl-webserver/pull/6 (so merge after that one if you want it I suppose).

Adds a multistage build, meaning that the image that runs the webapp is 54MB compressed as opposed to 500MB. The dev dependencies, caches and some other bloat are not in the final image in this one.

You can still build the dev image using:
```
docker build --target dev -t mauritso/ytdl-webserver-dev .
```

run only image using:
```
docker build -t mauritso/ytdl-webserver .
```

on the docker hub [here](https://hub.docker.com/r/mauritso/ytdl-webserver/tags/)

you can  try the image using:
```
docker run -p 3000:3000 mauritso/ytdl-webserver
```

### CODE REVIEW

**Comment by Algram** [comment]
Very nice, will definitely merge that afterwards :+1:

**Comment by mauritsvdvijgh** [comment]
btw you can also make the dev image a lot smaller and also use alpine. For that you would need to upgrade node-sass to 4.1 or higher for alpine support.

https://hub.docker.com/r/mauritso/ytdl-webserver-dev/tags/ example dev image.
Dockerfile: https://gist.github.com/mauritsvdvijgh/2d45b88e6599eba3e5ff564b434e80b1
Only works if you update node-sass to higher than 4.1 (the above image has 4.9 and seems to work fine)

**Comment by Algram** [comment]
@mauritsvdvijgh I'd really like to have the alpine image instead of ubuntu. I'll test the new node-sass locally and if there are no breaking changes, could you supply the updated Dockerfile in this PR?

Thanks!

**Comment by mauritsvdvijgh** [comment]
Done, no other changes except for the Dockerfile so it will not work in its current state (because of the older node-sass). After you tested/upgrade node-sass this should work.

**Comment by austinpray** [comment]
Totally forgot about multistage haha

File: Dockerfile [hunk 1]
```diff
@@ -1,29 +1,55 @@
...
+    rm -rf /root/.cache /var/cache /usr/share/terminfo  && \
+    ln /usr/bin/python3.6 /usr/bin/python               && \
+    mkdir -p public/temp
 
-# This is on a separate line because youtube-dl needs to be frequently updated
-RUN apt-get update \
-    && apt-get install -y youtube-dl \
-    && rm -rf /var/lib/apt/lists/*
+# install the latest npm and clear the cache
+RUN npm install -g npm@latest && \
```
Comment (austinpray) [inline Dockerfile]: might as well just merge this with above RUN directive to save on layers, right?

File: Dockerfile [hunk 2]
```diff
@@ -1,29 +1,55 @@
...
-COPY . ./
-RUN mkdir -p public/temp \
-    && npm run build
 
-EXPOSE 3000
-CMD [ "npm", "start" ]
+FROM base as dev
+
+# install all dependencies including dev and clear the cache
+RUN npm ci && \
```
Comment (austinpray) [inline Dockerfile]: Don't we want to copy over package-lock.json to bust this layer every time it changes?

**Comment by austinpray** [comment]
@mauritsvdvijgh haha yep I deleted the package-lock comment like 5 seconds after writing it. When I checked this branch out locally and saw the non-diff view it was right in front of my face.

Btw was just playing around to try to get node-sass to work and stuff like this doesn't work:

```diff
diff --git a/Dockerfile b/Dockerfile
index a99371f..bd3a58b 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -4,11 +4,14 @@ WORKDIR /home/app
 EXPOSE 3000
 CMD [ "npm", "start" ]
 
+ENV SASS_BINARY_DIR /usr/lib
+
 RUN apk add --update --no-cache \
         python3 \
         ffmpeg \
         nodejs \
         nodejs-npm \
+        libsass \
         ca-certificates && \
     find / -type d -name __pycache__ -exec rm -r {} +   && \
     rm -r /usr/lib/python*/ensurepip                    && \

```

Just gotta use node-sass@4.9.3 I guess

**Comment by Algram** [comment]
Fixed by https://github.com/Algram/ytdl-webserver/commit/473155a7c65075e21bf2e5c501227f5f63b11962

**Pull request closed without merging**